go/hyperkit.go

```diff
@@ -0,0 +1,526 @@
+		}
+	}
+	h.process, err = os.FindProcess(h.Pid)
+	if err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+// SetLogger sets the log instance to use for the output of the hyperkit process itself (not the console of the VM).
+// This is only relevant when Console is set to ConsoleFile
+func (h *HyperKit) SetLogger(logger *log.Logger) {
+	h.log = logger
+}
+
+// Run the VM with a given command line until it exits
+func (h *HyperKit) Run(cmdline string) error {
+	h.background = false
+	return h.execute(cmdline)
+}
+
+// Start the VM with a given command line in the background
+func (h *HyperKit) Start(cmdline string) error {
+	h.background = true
+	return h.execute(cmdline)
+}
+
+func (h *HyperKit) execute(cmdline string) error {
+	var err error
+	// Sanity checks on configuration
+	if h.Console == ConsoleFile && h.StateDir == "" {
+		return fmt.Errorf("If ConsoleFile is set, StateDir was be specified")
+	}
+	if h.UserData != "" && h.ISOImage != "" {
+		return fmt.Errorf("If UserData is supplied, ISOImage must not be set")
+	}
+	if h.ISOImage != "" {
+		if _, err = os.Stat(h.ISOImage); os.IsNotExist(err) {
+			return fmt.Errorf("ISO %s does not exist", h.ISOImage)
+		}
+	}
+	if h.UserData != "" && h.StateDir == "" {
+		return fmt.Errorf("If UserData is supplied, StateDir was be specified")
+	}
+	if _, err = os.Stat(h.Kernel); os.IsNotExist(err) {
+		return fmt.Errorf("Kernel %s does not exist", h.Kernel)
+	}
+	if _, err = os.Stat(h.Initrd); os.IsNotExist(err) {
+		return fmt.Errorf("initrd %s does not exist", h.Initrd)
+	}
+	if h.DiskImage == "" && h.StateDir == "" && h.DiskSize != 0 {
+		return fmt.Errorf("Can't create disk, because neither DiskImage nor StateDir is set")
+	}
+
+	// Create files
+	if h.StateDir != "" {
+		err = os.MkdirAll(h.StateDir, 0755)
+		if err != nil {
+			return err
+		}
+	}
+	if h.DiskImage == "" && h.DiskSize != 0 {
+		h.DiskImage = filepath.Join(h.StateDir, "disk.img")
+	}
+	if _, err = os.Stat(h.DiskImage); os.IsNotExist(err) {
+		if h.DiskSize != 0 {
+			err = CreateDiskImage(h.DiskImage, h.DiskSize)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if h.UserData != "" {
+		h.ISOImage, err = createUserDataISO(h.StateDir, h.UserData)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Run
+	h.buildArgs(cmdline)
+	err = h.execHyperKit()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Stop the running VM
+func (h *HyperKit) Stop() error {
+	if h.process == nil {
+		return fmt.Errorf("hyperkit process not known")
+	}
+	if !h.IsRunning() {
+		return nil
+	}
+	err := h.process.Kill()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsRunning returns true if the hyperkit process is running.
+func (h *HyperKit) IsRunning() bool {
+	// os.FindProcess on Unix always returns a process object even
+	// if the process does not exists. There does not seem to be
+	// a call to find out if the process is running either, so we
+	// use another package to find out.
+	proc, err := ps.FindProcess(h.Pid)
```

Comment:
Oh wait, this handles the case where it has died since you last looked at it. You can't use `Wait` as a trigger in the case where you have attached to an existing process, so this is as good as it gets...